auth/user_mgt.go

```diff
@@ -0,0 +1,711 @@
+//
+// Custom claims set via this function can be used to define user roles and privilege levels.
+// These claims propagate to all the devices where the user is already signed in (after token
+// expiration or when token refresh is forced), and next time the user signs in. The claims
+// can be accessed via the user's ID token JWT. If a reserved OIDC claim is specified (sub, iat,
+// iss, etc), an error is thrown. Claims payload must also not be larger then 1000 characters
+// when serialized into a JSON string.
+func (c *Client) SetCustomUserClaims(ctx context.Context, uid string, customClaims map[string]interface{}) error {
+	if customClaims == nil || len(customClaims) == 0 {
+		customClaims = map[string]interface{}{}
+	}
+	_, err := c.UpdateUser(ctx, uid, (&UserToUpdate{}).CustomClaims(customClaims))
+	if err != nil {
+		return err
+	}
+	return err
+}
+
+// ------------------------------------------------------------
+// Setters and utilities for Create and Update input structs.
+
+func (p *commonParams) appendErrString(s string, i ...interface{}) {
+	p.errors = append(p.errors, fmt.Sprintf(s, i...))
+}
+
+func (p *commonParams) payloadInitialized() {
+	if p.payload == nil {
+		p.payload = make(map[string]interface{})
+	}
+}
+
+func (p *commonParams) set(key string, value interface{}) {
+	p.payloadInitialized()
+	p.payload[key] = value
+}
+
+// assumes that payloadName is a string field in p.payload
+func processDeletion(p *commonParams, payloadName, errorName string) {
+	if dn, ok := p.payload[payloadName]; ok && len(dn.(string)) == 0 {
+		p.addToListParam(deletionSpecs[payloadName].deleteListName, deletionSpecs[payloadName].deleteFieldName)
+		delete(p.payload, payloadName)
+	}
+}
+
+func processClaims(p *commonParams, payloadName, errorName string) {
+	if _, ok := p.payload[payloadName]; !ok {
+		return
+	}
+	p.checkReservedClaims()
+	p.setCustomAttributes()
+}
+func (p *commonParams) checkReservedClaims() {
+	cc := p.payload["customClaims"]
+	switch claims := cc.(type) {
+	case map[string]interface{}:
+		for _, key := range reservedClaims {
+			if _, ok := claims[key]; ok {
+				p.appendErrString("claim %q is reserved, and must not be set", key)
+			}
+		}
+	}
+}
+
+func (p *commonParams) setCustomAttributes() {
+	cc := p.payload["customClaims"]
+	b, err := json.Marshal(cc)
+	if err != nil {
+		p.appendErrString("invalid custom claims Marshaling error: %v wanted %v", err, cc)
+		return
+	} //else if len(b) > maxLenPayloadCC {
+	//		p.appendErrString("Custom Claims payload must not exceed %d characters", maxLenPayloadCC)
+	//	} else {
+	s := string(b)
+	if s == "null" {
+		s = "{}"
+	}
+}
+
+func (p *commonParams) addToListParam(listname, param string) {
+	if _, ok := p.payload[listname]; ok {
+		p.payload[listname] = append(p.payload[listname].([]string), param)
+	} else {
+		p.set(listname, []string{param})
+	}
+}
+
+// Validators.
+
+// No validation needed. used for bool fields
+func allowed(p *commonParams, _, _ string) {
```

Comment:
I would expect all these validators to return an error. The caller should fail-fast on the first error encountered.

Also follow a consistent naming scheme for all validator functions -- e.g. validateEmail, validatePhone, validateDisplayName etc